src/handleSubmit.js

```diff
@@ -9,88 +10,86 @@ const handleSubmit = (submit, props, valid, asyncValidate, fields) => {
 
   touch(...fields) // mark all fields as touched
 
+  let doStopSubmit = () => {}
+  const doStartSubmit = () => {
+    doStopSubmit = stopSubmit
+    startSubmit()
+  }
+  
   // XXX: Always submitting when persistentSubmitErrors is enabled ignores sync errors.
   // It would be better to check whether the form as any other errors except submit errors.
   // This would either require changing the meaning of `valid` (maybe breaking change),
   // having a more complex conditional in here, or executing sync validation in here
   // the same way as async validation is executed.
-  if (valid || persistentSubmitErrors) {
-    const doSubmit = () => {
-      let result
-      try {
-        result = submit(values, dispatch, props)
-      } catch (submitError) {
-        const error = submitError instanceof SubmissionError ? submitError.errors : undefined
-        setSubmitFailed(...fields)
-        if (onSubmitFail) {
-          onSubmitFail(error, dispatch)
-        }
-        if (error || onSubmitFail) {
-          // if you've provided an onSubmitFail callback, don't re-throw the error
-          return error
-        } else {
-          throw submitError
-        }
+
+  let asyncValidationFailed = false
```

Comment:
Depending on whether the async validation fails or not the errors have to be returned in a different way (some tests check for that). I think handleSubmit should always return the errors in the same way.